### DISCUSSION TITLE
테스트 코드 Fixture 및 Display 컨벤션 건의 사항

### DISCUSSION THREAD

**Discussion by KIMSEI1124** [What-s-Your-Plan/wypl-backend#52] [Ideas]
# 1. Fixture 컨벤션

Fixture를 정의할 때 가독성을 높이기 위해 한글로 변수 이름과 메서드 이름을 작성합니다. 이는 테스트 코드 작성 시 이해하기 쉽도록 하며, UTF-8 환경에서 문제없이 사용할 수 있습니다.

## 코드 예시

```java
public enum CalendarFixture {
    와플_프로젝트("와플 개발 일정", "와플 스터디 개발 일정 달력입니다.", 1L);

    private final String name;
    private final String description;
    private final long ownerId;

    CalendarFixture(String name, String description, long ownerId) {
        this.name = name;
        this.description = description;
        this.ownerId = ownerId;
    }

    public Calendar 공유달력을_생성한다() {
        return Calendar.builder()
            .name(name)
            .description(description)
            .ownerId(ownerId)
            .isShared(true)
            .build();
    }

    public Calendar 비공유달력을_생성한다() {
        return Calendar.builder()
            .name(name)
            .description(description)
            .ownerId(ownerId)
            .isShared(false)
            .build();
    }
}
```

## 특징

- 한글 사용: 변수와 메서드 이름을 한글로 작성하여 직관적으로 이해할 수 있습니다.
- 메서드명 분리: 공유 달력과 비공유 달력을 생성하는 메서드를 분리하여 명확히 구분합니다.

## Fixture 사용 예시

```java
@BeforeEach
public void setUp() {
    calendar = 와플_프로젝트.공유달력을_생성한다();
    repository.save(calendar);
}
```

# 2. DisplayName 컨벤션

테스트 코드의 `@DisplayName`은 테스트 대상 도메인의 엔티티 이름을 명확히 나타내는 것을 권장합니다.

## 코드 예시

```java
@DisplayName("**Calendar** 조회에 성공한다.") // 달력 조회에 성공한다.
@Test
void success() {
    /* When & Then */
    assertThatCode(() -> CalendarRepositoryUtils.findById(repository, calendar.getId()))
        .doesNotThrowAnyException();
}
```

## 특징

- 도메인 명시: 테스트 대상 도메인(예: Calendar)을 `@DisplayName`에 포함하여 테스트의 목적을 명확히 합니다.
- 도메인 중심 표현: “달력 조회에 성공한다” 대신 “Calendar 조회에 성공한다”와 같이 도메인 중심적으로 표현합니다.

# 3. 기대 효과

테스트 코드 가독성 향상: 한글로 작성된 Fixture와 메서드명을 통해 테스트 코드의 가독성이 높아지고, 의도를 명확히 전달할 수 있습니다.

도메인 명확성: `@DisplayName`에 도메인을 명시함으로써 테스트 코드가 어떤 엔티티를 테스트하고 있는지 명확히 알 수 있습니다.

# 4. 적용 사례

```java
@BeforeEach
public void setUp() {
    calendar = 와플_프로젝트.공유달력을_생성한다();
    repository.save(calendar);
}

@DisplayName("Calendar를 저장하고 조회에 성공한다.")
@Test
void saveAndFindCalendar() {
    /* When */
    Calendar foundCalendar = CalendarRepositoryUtils.findById(repository, calendar.getId());

    /* Then */
    assertThat(foundCalendar).isNotNull();
    assertThat(foundCalendar.getName()).isEqualTo(calendar.getName());
}
```

이와 같은 컨벤션을 따르면 테스트의 의도를 명확히 표현할 수 있습니다.

**Comment by sjhjack** [comment]
좋은 의견 감사합니다.
명확한 의도 전달 측면에서 장점이 있어 보입니다.
다만, 다른 의견이 있어서 코멘트 남깁니다.

저는 `공유달력을_생성한다` 라는 한글 메서드명 보다는 영어가 낫다고 생각합니다.
위 메서드 명을 저의 경우 `createSharedCalendar` 와 같이 중심이 되는 변수명을 포함해서 작성하는 방법을 선호합니다.
이는 다루고자 하는 대상을 명확히하여 개발 시 혼동이 적다는 장점이 있습니다.

또한, Display 컨벤션으로 도메인 중심적 표현을 사용하는 것과 Fixture의 메서드 명을 한글로 작성하는 것은 일관적이지 않다고 생각합니다.
메서드 명도 약속된 언어인 영어로 작성하는 것이 의도를 명확히 하기에 좋을 것 같습니다.

**Reply by KIMSEI1124** [reply]
좋은 의견 감사합니다! 👍🏼 

그래서 현재 코드는 영어로 작성하였습니다! 아래는 실제 작성한 코드입니다! 한번 봐주시고 문제점이 있다면 말씀해주시면 감사합니다!

```java
@DisplayName("Member 조회에 성공한다.")
@ParameterizedTest
@EnumSource(MemberFixture.class)
void successTest(MemberFixture fixture) {
	/* Given */
	given(repository.findById(anyLong()))
		.willReturn(Optional.of(fixture.toMember()));

	/* When & Then */
	assertThatCode(() -> MemberRepositoryUtils.findById(repository, memberId))
		.doesNotThrowAnyException();
}
```

- `toMember()` 을 사용하여 객체를 생성하였습니다!
- Display에서 Entity를 작성할때 클래스 이름으로 직접 작성하여 강조하였습니다! "**Member** 조회에 성공한다."
